lib/public_key/src/pubkey_policy_tree.erl

```diff
@@ -0,0 +1,421 @@
+%%
+%% That is this function caclualates the valid policy node
+%% set after being constrained either by the user or only by the
+%% authorities (certificate extensions).
+%% --------------------------------------------------------------------
+constraint_policy_node_set({}) ->
+    [];
+constraint_policy_node_set(Tree) ->
+    case any_leaves(Tree) of
+        [] ->
+            constrain(Tree);
+        AnyLeaves ->
+            AnyLeaves
+    end.
+
+%%--------------------------------------------------------------------
+-spec empty() -> policy_tree().
+%%--------------------------------------------------------------------
+empty() ->
+    {}.
+
+%%--------------------------------------------------------------------
+-spec in_set(Policy::public_key:oid(), Set::[policy_node()]) -> boolean().
+%%
+%% Is there a node with <Policy> in <Set>
+%%--------------------------------------------------------------------
+in_set(_, []) ->
+    false;
+in_set(Policy, [#{valid_policy := Policy} |_]) ->
+    true;
+in_set(Policy, [_ | Rest]) ->
+    in_set(Policy, Rest).
+
+%%--------------------------------------------------------------------
+-spec is_empty(policy_tree()) -> boolean().
+%%--------------------------------------------------------------------
+is_empty({}) ->
+    true;
+is_empty(_) ->
+    false.
+
+%%--------------------------------------------------------------------
+-spec map_leaves(policy_tree(), LeafFun::function()) -> policy_tree().
+%%
+%% Update all leaves as determined by <LeafFun>
+%%--------------------------------------------------------------------
+map_leaves({Parent, [{_, _}|_] = SubTree}, LeafFun) ->
+    {Parent, lists:map(fun(ChildNode)->
+                               map_leaves(ChildNode, LeafFun)
+                       end, SubTree)};
+map_leaves({Parent, Leaves}, LeafFun) ->
+    {Parent, lists:map(LeafFun, Leaves)}.
+
+%%--------------------------------------------------------------------
+-spec prune_leaves(policy_tree(), Policy::public_key:oid()) -> policy_tree().
+%%
+%% Delete all leaves with the valid_policy <Policy>
+%%--------------------------------------------------------------------
+prune_leaves({} = Empty, _) ->
+    Empty;
+prune_leaves({_, _} = Tree, Policy) ->
+    LeafFun = fun(#{valid_policy := ValidPolicy} = Node) ->
+                      case ValidPolicy of
+                          Policy ->
+                              false;
+                          _ ->
+                              {true, Node}
+                      end
+              end,
+    filter_leaves(Tree, LeafFun).
+
+%%--------------------------------------------------------------------
+-spec prune_tree(policy_tree()) -> policy_tree().
+%%
+%% Delete branches that are shorher than the total depth of the tree.
+%%--------------------------------------------------------------------
+prune_tree({} = Empty) ->
+    Empty;
+prune_tree({_, []}) ->
+    empty();
+prune_tree({Root, SubTree}) ->
```

Comment:
tree below will be returned not pruned. 
is this expected?

```
{#{valid_policy => {2,5,29,32,0},
   expected_policy_set => [{2,5,29,32,0}],
   qualifier_set => []},
 [{#{valid_policy => "GOLD",
     expected_policy_set => ["GOLD","SILVER"],
     qualifier_set => []},
   [#{valid_policy => "GOLD",
      expected_policy_set => ["GOLD"],
      qualifier_set => []}]},
  #{valid_policy => "SILVER",
    expected_policy_set => ["GOLD","SILVER"],
    qualifier_set => []}]}
```
![image](https://github.com/erlang/otp/assets/4620572/8c9d18fe-e760-496b-a33b-385bcb5e941a)

it prunes tree, if leaf of too short branch is of policy_node_tree type with empty children list. but does not if this leaf is of policy_node type.

Reply:
At pruning time trees will not look like that as add leaves will make sure  that is not the case.

Reply:
Example:
 After adding leafs to next level:
 ```
{#{expected_policy_set => [{2,5,29,32,0}],
                     valid_policy => {2,5,29,32,0},
                     qualifier_set => []},
                   [{#{expected_policy_set => [{2,16,840,1,101,3,2,1,48,2}],
                       valid_policy => {2,16,840,1,101,3,2,1,48,2},
                       qualifier_set => []},
                     []},
                    {#{expected_policy_set => [{2,16,840,1,101,3,2,1,48,1}],
                       valid_policy => {2,16,840,1,101,3,2,1,48,1},
                       qualifier_set => []},
                     [#{expected_policy_set => [{2,16,840,1,101,3,2,1,48,1}],
                        valid_policy => {2,16,840,1,101,3,2,1,48,1},
                        qualifier_set => []}]}]}
```
after pruning:
 ```
{#{expected_policy_set => [{2,5,29,32,0}],
                    valid_policy => {2,5,29,32,0},
                    qualifier_set => []},
                  [{#{expected_policy_set => [{2,16,840,1,101,3,2,1,48,1}],
                      valid_policy => {2,16,840,1,101,3,2,1,48,1},
                      qualifier_set => []},
                    [#{expected_policy_set => [{2,16,840,1,101,3,2,1,48,1}],
                       valid_policy => {2,16,840,1,101,3,2,1,48,1},
                       qualifier_set => []}]}]}


```

Reply:
@u3s check and resolve if OK

Reply:
convinced. I've updated test case.